firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+import androidx.core.app.FrameMetricsAggregator;
+import androidx.fragment.app.Fragment;
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Optional;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ *
+ * <p>IMPORTANT: each recorder holds a reference to an Activity, so it is very important to
+ * dereference each recorder at or before its Activity's onDestroy. Similar for Fragments.
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+
+  private final Activity activity;
+  private final FrameMetricsAggregator frameMetricsAggregator;
+  private final Map<Fragment, PerfFrameMetrics> subTraceMap;
+
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity.
+   *
+   * @param activity the activity that the recorder is collecting data from.
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new HashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      Map<Fragment, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.frameMetricsAggregator = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window. */
+  public void start() {
+    if (isRecording) {
```

Comment:
I am not sure what value that API would add because I don't see us using it. We are just going to call `start()` in `onActivityStart` and `stop()` in `onActivityStop`, if it fails then we are not doing a symmetrical call and are calling `start()` somewhere else, which is wrong and it should output warning.

Reply:
+1 to Leo

Reply:
I'm mixed on this. I have seen usecases where having such an API would be very useful. But in the interest of self-containment, I'm fine to move ahead here.